main/src/ca/uwaterloo/flix/language/phase/Safety.scala

```diff
@@ -82,115 +86,138 @@ object Safety {
-      visitExp(exp1)(inTryCatch = false, renv, flix) ++ visitExp(exp2)
+      visitExp(exp1)(inTryCatch = false, renv, flix, sctx)
+      visitExp(exp2)
 
     case Expr.Region(_, _) =>
-      Nil
+      ()
 
     case Expr.Scope(_, _, exp, _, _, _) =>
       visitExp(exp)
 
     case Expr.IfThenElse(exp1, exp2, exp3, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
+      visitExp(exp1)
+      visitExp(exp2)
+      visitExp(exp3)
 
     case Expr.Stm(exp1, exp2, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2)
+      visitExp(exp1)
+      visitExp(exp2)
 
     case Expr.Discard(exp, _, _) =>
       visitExp(exp)
 
     case Expr.Match(exp, rules, _, _, _) =>
-      visitExp(exp) ++ rules.flatMap(rule => rule.guard.map(visitExp).getOrElse(Nil) ++ visitExp(rule.exp))
+      visitExp(exp)
+      rules.foreach { rule =>
+        rule.guard.foreach(visitExp)
+        visitExp(rule.exp)
+      }
 
     case Expr.TypeMatch(exp, rules, _, _, _) =>
       // check whether the last case in the type match looks like `..: _`
-      val missingDefault = rules.lastOption match {
+      rules.lastOption match {
         // Use top scope since the rigidity check only cares if it's a syntactically known variable
         case Some(TypeMatchRule(_, Type.Var(sym, _), _)) if renv.isFlexible(sym)(Scope.Top) =>
-          Nil
+          ()
         case Some(_) | None =>
-          List(SafetyError.MissingDefaultTypeMatchCase(exp.loc))
+          sctx.errors.add(SafetyError.MissingDefaultTypeMatchCase(exp.loc))
       }
-      visitExp(exp) ++ missingDefault ++ rules.flatMap(rule => visitExp(rule.exp))
+      visitExp(exp)
+      rules.foreach(rule => visitExp(rule.exp))
 
     case Expr.RestrictableChoose(_, exp, rules, _, _, _) =>
-      visitExp(exp) ++ rules.flatMap(rule => visitExp(rule.exp))
+      visitExp(exp)
+      rules.foreach(rule => visitExp(rule.exp))
 
       case Expr.Tag(_, exps, _, _, _) =>
-        exps.flatMap(visitExp)
+        exps.foreach(visitExp)
 
       case Expr.RestrictableTag(_, exps, _, _, _) =>
-        exps.flatMap(visitExp)
+        exps.foreach(visitExp)
 
     case Expr.Tuple(elms, _, _, _) =>
-      elms.flatMap(visitExp)
+      elms.foreach(visitExp)
 
     case Expr.RecordEmpty(_, _) =>
-      Nil
+      ()
 
     case Expr.RecordSelect(exp, _, _, _, _) =>
       visitExp(exp)
 
     case Expr.RecordExtend(_, value, rest, _, _, _) =>
-      visitExp(value) ++ visitExp(rest)
+      visitExp(value)
+      visitExp(rest)
 
     case Expr.RecordRestrict(_, rest, _, _, _) =>
       visitExp(rest)
 
     case Expr.ArrayLit(elms, exp, _, _, _) =>
-      elms.flatMap(visitExp) ++ visitExp(exp)
+      elms.foreach(visitExp)
+      visitExp(exp)
 
     case Expr.ArrayNew(exp1, exp2, exp3, _, _, _) =>
-      visitExp(exp1) ++ visitExp(exp2) ++ visitExp(exp3)
+      visitExp(exp1)
+      visitExp(exp2)
+      visitExp(exp3)
 
     case Expr.ArrayLoad(base, index, _, _, _) =>
```

Comment:
This has nothing to do with your pr, but if you could rename base -> exp1, index -> exp2.